src/benchmarks/src/rpc_service_bench.cpp

```diff
@@ -0,0 +1,286 @@
+      log_error(LOGGING_TAG_RPC_BENCH, "RPCBench failed to connect to redis",
+                "Retrying in 1 second...");
+      std::this_thread::sleep_for(std::chrono::seconds(1));
+    }
+    while (!redis_subscriber_.connect(conf.get_redis_address(),
+                                      conf.get_redis_port())) {
+      log_error(LOGGING_TAG_RPC_BENCH,
+                "RPCBench subscriber failed to connect to redis",
+                "Retrying in 1 second...");
+      std::this_thread::sleep_for(std::chrono::seconds(1));
+    }
+    redis::send_cmd_no_reply<std::string>(
+        redis_connection_, {"CONFIG", "SET", "notify-keyspace-events", "AKE"});
+    redis::subscribe_to_container_changes(
+        redis_subscriber_,
+        // event_type corresponds to one of the Redis event types
+        // documented in https://redis.io/topics/notifications.
+        [this](const std::string &key, const std::string &event_type) {
+          if (event_type == "hset") {
+            auto container_info =
+                redis::get_container_by_key(redis_connection_, key);
+            benchmark_container_id_ =
+                std::stoi(container_info["zmq_connection_id"]);
+
+            // SEND FIRST MESSAGE
+            send_message();
+          }
+
+        });
+  }
+
+  Benchmarker(const Benchmarker &other) = delete;
+  Benchmarker &operator=(const Benchmarker &other) = delete;
+
+  Benchmarker(Benchmarker &&other) = default;
+  Benchmarker &operator=(Benchmarker &&other) = default;
+  ~Benchmarker() {
+    std::unique_lock<std::mutex> l(bench_completed_cv_mutex);
+    bench_completed_cv_.wait(l, [this]() { return bench_completed_ == true; });
+  }
+
+  void send_message() {
+    cur_msg_start_time_millis_ =
+        std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::system_clock::now().time_since_epoch())
+            .count();
+    cur_message_id_ =
+        rpc_->send_message(request_.serialize(), benchmark_container_id_);
+  }
+
+  void on_response_recv(rpc::RPCResponse response) {
+    // process the response
+    long recv_time_millis =
+        std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::system_clock::now().time_since_epoch())
+            .count();
+    if (response.first != cur_message_id_) {
+      std::stringstream ss;
+      ss << "Response message ID ";
+      ss << response.first;
+      ss << " did not match in flight message ID ";
+      ss << cur_message_id_;
+      throw std::logic_error(ss.str());
+    }
+    long message_duration_millis =
+        recv_time_millis - cur_msg_start_time_millis_;
+    std::vector<float> deserialized_outputs =
+        deserialize_outputs(response.second);
+    float sum = 0.0;
+    for (float f : deserialized_outputs) {
+      sum += f;
+    }
+    log_info_formatted(LOGGING_TAG_RPC_BENCH, "output sum: {}", sum);
+    msg_latency_hist_->insert(message_duration_millis);
+    throughput_meter_->mark(1);
```

Comment:
So that's actually how the containers work. Each container can only handle one message at a time. So this should reflect the same throughput as the full system. Keep in mind that we're sending mini-batches though, so each message contains many requests.